HSEARCH-2254 Support (single-valued) sorts on fields within nested fields

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-2254

fax4ever commented:
Still in preview: I handled only Field sort.
I think I have to handle at least the Distance one too.

I would prefer have a review on what I've done so far, before continue. Since the issue is not trivial at all.

To be honest, it was very hard for me
and I'm pretty sure we could improve several choices I made here.
For instance we will avoid to run a top doc collector again just to load the nested contributions.

Looking at Elasticsearch's code it seems that they load the nested document using
not `ToParentBlockJoinQuery` nor `ToChildBlockJoinQuery` but `ParentChildrenBlockJoinQuery`.
With the latter we need to pass the parent id. So we would have needed to run one for each parent document.
For that reason here I opted for the ToChildBlockJoinQuery`, that's same as we used for supporting projection on nested fields.

This issue is more difficult than the one on projection. Here we cannot simply merge all contributions. We need to switch document id ( parent to nested ) for each nested field ( so the switching and the convertion map are scoped on the single field ), paying attention do not alter what we already do on flattened or root object fields: see `CompositeSortIT`.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneTextFieldSortBuilder.java:
```diff
@@ -36,7 +37,14 @@ protected Object encodeMissingAs(F converted) {
 	public void buildAndContribute(LuceneSearchSortCollector collector) {
 		// For STRING type, missing value must be either STRING_FIRST or STRING_LAST.
 		// Otherwise we need a CUSTOM type.
```
You should probably move this comment to `getSortField()`

fax4ever replied:
I squashed both to `HSEARCH-2254 Support sort on nested field - Lucene`.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneTextFieldSortBuilder.java:
```diff
@@ -36,7 +37,14 @@ protected Object encodeMissingAs(F converted) {
+		collector.collectSortField( getSortField(), nestedDocumentPath );
+	}
+
+	private SortField getSortField() {
```
Maybe rather `createSortField()`? It's always confusing to have a getter that doesn't just "get". Also, you're going to add side-effects to this getter in later commits.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/impl/LuceneNestedDocumentsSort.java:
```diff
@@ -32,4 +47,50 @@ public void add(LuceneNestedDocumentFieldContribution contribution) {
+	public void processNestedPaths(Query luceneQuery, IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		for ( Map.Entry<String, Set<LuceneNestedDocumentFieldContribution>> fieldEntry : contributions.entrySet() ) {
+			String nestedDocumentPath = fieldEntry.getKey();
+			nestedDocumentMaps.put( nestedDocumentPath, getFetchNestedDocumentMap( nestedDocumentPath, luceneQuery, indexSearcher, scoreDocs ) );
```
So I guess you intend to improve on this, since one query + one query per sort doesn't look too good. Two I can understand, but 1 + N is a bit extreme.

Can't the `LuceneChildrenCollector` be adapted to produce multiple maps? Whenever it inspects a document, it could inspect its "nested document path" and decide to which map it should go based on that.

fax4ever replied:
Currently we do one query for each nested document paths that is present on some sort field. The same result is reused among all fields on the same nested object. Actually I don't know if there is a way to improve that. Since different nested document paths can have different nested documents. If there was, I would happy to improve that.

fax4ever replied:
From what I've got Elasticsearch does one ParentChildrenBlockJoinQuery for each parent document. That should be definitely worse.

yrodiere replied:
I'm not sure we're talking about the same thing. As far as I can tell, `org.elasticsearch.search.sort.SortBuilder#resolveNestedQuery` is used to produce the query that will return all nested documents, 

You're right that they don't seem to attempt to run one query for everything, though: apparently they just run `org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested#innerDocs` once for each sort, which is similar to what you're doing.

I suspect their use of a DocIdSet is much more efficient than the Map you're using, though. In particular they seem to retrieve nested documents... on the fly!

* `org.elasticsearch.search.sort.SortBuilder#resolveNested(org.elasticsearch.index.query.QueryShardContext, org.elasticsearch.search.sort.NestedSortBuilder)` creates, for a given query, a given sort and a given list of nested paths, a `Nested` utility object. I think it's similar to your `LuceneNestedDocumentFieldContribution`, except it has knowledge of the query.
* `org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested#innerDocs` returns, for a given `LeafReaderContext` (~a segment of the index), a `DocIdSetIterator` that will return all relevant nested documents.
* `org.apache.lucene.search.FieldComparator.TermValComparator#getBinaryDocValues` (and similar classes for other types of DocValues) wraps the DocValues to be used for sorts with a wrapper that uses `MultiValueMode#select`. It assumes these wrapped DocValues will transparently redirect to nested documents when the value of the parent document is requested.
* `org.elasticsearch.search.MultiValueMode#select(org.elasticsearch.index.fielddata.SortedBinaryDocValues, org.apache.lucene.util.BytesRef, org.apache.lucene.util.BitSet, org.apache.lucene.search.DocIdSetIterator, int, int)` (and similar methods for other types of DocValues)  takes advantage of the nested documents "docIdSet" to make sure that whenever someone requests the docvalues to advance to parent document, the doc values will instead advance to the relevant nested document if necessary.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/impl/LuceneNestedFieldComparator.java:
```diff
@@ -18,4 +18,21 @@
+		}
+
+		// TODO at the moment we do not support multi value sorting on nested document
+		// using one of them in case
```
I think you could use a `Map<Integer, Integer>`, in that case. I suspect that, even when we'll support multi-value sorting, we'll do the nested document selection upfront, not in here, so we'll never need more than one document ID here.
That would spare us one `Set` instantiation per document, which may or may not be a big deal, but I suspect it will.

fax4ever replied:
I thought about it and I think you definitely right. Given a single nested document path we can have at most one nested document for each root document.
I think that we have a set here, just because I've reused the same nested query we already had for nested projection.

fax4ever replied:
Done in a different commit. We could revert it, if we were to realize that we were wrong.

fax4ever commented:
Thanks @yrodiere. Rebased the changes.
Now I can address your comments.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/work/impl/LuceneSearchWork.java:
```diff
@@ -38,13 +40,14 @@
 
 	LuceneSearchWork(Query luceneQuery,
-			Sort luceneSort,
+			Sort luceneSort, Set<String> nestedPathsInSort,
```
You were right. In fact I put the nested sort information in `LuceneSearchQueryRequestContext` when I did the rebase.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/LuceneCollectors.java:
```diff
@@ -60,10 +75,12 @@ public void collect(IndexSearcher indexSearcher, Query luceneQuery, int offset,
+			handleRescoring( indexSearcher, luceneQuery );
+		}
+		if ( nestedPathsInSort != null && !nestedPathsInSort.isEmpty() ) {
+			extractTopDocsUsingTheirNested( indexSearcher, luceneQuery, offset, limit );
```
I moved `extractTopDocsUsingTheirNested` above.
But without other changes we need to call `extractTopDocs` before we call `extractTopDocsUsingTheirNested`, since we need the `ScoreDoc[] scoreDocs` to f etch the `nestedDocumentMap`.

yrodiere replied:
Ok... now that I think about it, I think there's a flaw in this approach.
What about the following scenario:

* The query matches documents A, B, C
* The limit is 2
* The first search returns top documents [`A, B]`
* The second search, taking into account nested documents, would have returned `[C, B]` (`A` excluded because it's third), but since `C` was not in the top docs of the first search, it ends up returning `[B, A]`.

I suspect this is a very real possibility...

The only option to avoid this problem would be to revert the logic:

* Currently you perform the `search()` call once, and then if you discover that there are nested sorts, you try to "fix" the first results by running a second `search()`.
* The alternative would be to detect *before any `search()`* that there are nested sorts. In that case you would run a preliminary `search()` to collect nested documents (and only nested documents) *without a limit*. Then you would run the actual `search()` with a limit.

I'm aware this would be require an awful lot of memory for big indexes, but at least it would work (?). We can create a ticket to try and optimize this, maybe.

yrodiere replied:
Actually, I've had a look at the Elasticsearch code, and it seems they retrieve nested documents on the fly when it comes to sorts. See my other comment. I think if you manage to implement a solution that does the same, the flaw I just mentioned will just disappear. Could you have a look, please?

fax4ever replied:
Sure

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/LuceneCollectors.java:
```diff
@@ -78,6 +95,42 @@ public TopDocs getTopDocs() {
+		fillNestedDocumentMap( indexSearcher, childrenCollector );
+		LuceneTopNestedFieldCollector topNestedCollector = new LuceneTopNestedFieldCollector( sort, maxDocs, nestedDocumentMap );
+
+		indexSearcher.search( luceneQuery, topNestedCollector );
```
I removed this collector.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/LuceneTopNestedFieldCollector.java:
```diff
@@ -0,0 +1,98 @@
+		initReverseMap();
+	}
+
+	private void initReverseMap() {
```
We don't need it anymore

fax4ever commented:
Thanks @yrodiere. I think I've address all the comments except the one about to optimize nested queries when we have both nested sorts and nested projections.

I'm not sure we can optimize something, since projections and sorts can work on different nested document paths. Let's talk on Zulip.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneNumericFieldSortBuilder.java:
```diff
@@ -32,13 +34,25 @@
+		collector.collectSortField( getSortField(), nestedFieldContribution );
+	}
+
+	private SortField getSortField() {
```
Looks like your renamed this to `createSortField` in the text field sort builder, but not here.

fax4ever commented:
@yrodiere thanks!
I think that now we have an `on-the-fly` nested field sorting.
Have I missed some comments?

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/onthefly/doubleval/impl/SortedNumericDoubleValues.java:
```diff
@@ -0,0 +1,59 @@
+ * Copied and adapted from {@code org.elasticsearch.index.fielddata.SortedNumericDoubleValues} class
+ * of <a href="https://github.com/elastic/elasticsearch">Elasticsearch project</a>.
+ */
+public abstract class SortedNumericDoubleValues {
```
I think this, and all the classes it relies on, is not currently necessary, as we could work directly on longs all along.
Diffentiating longs and doubles only makes sense when handling sorts on multi-valued fields and sorting on the average (because I don't think we can compute an avergae directly on the long-encoded representation of doubles).

Anyway... let's keep it that way if you want, I guess it will be useful later.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/onthefly/doubleval/impl/NumericDoubleValues.java:
```diff
@@ -0,0 +1,112 @@
+	protected NumericDoubleValues() {
+	}
+
+	// TODO: this interaction with sort comparators is really ugly...
```
It was. Looking at it, it seems that even they are not very happy about the solution they adopt. :)

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/onthefly/doubleval/impl/NumericDoubleValues.java:
```diff
@@ -0,0 +1,112 @@
+ * Copied and adapted from {@code org.elasticsearch.index.fielddata.NumericDoubleValues} class
+ * of <a href="https://github.com/elastic/elasticsearch">Elasticsearch project</a>.
+ */
+public abstract class NumericDoubleValues extends DoubleValues {
```
Good idea. Let's rename it.

fax4ever commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/sort/CompositeSearchSortIT.java:
```diff
@@ -190,13 +190,77 @@ public void then() {
+	}
+
+	@Test
+	public void then_flattened_nested_limit2() {
```
I would like to keep it anyway.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/onthefly/impl/NestedNumericFieldComparatorSource.java:
```diff
@@ -0,0 +1,108 @@
+
+	@Override
+	public FieldComparator<?> newComparator(String fieldname, int numHits, int sortPos, boolean reversed) {
+		switch ( type ) {
```
you definitely right

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/onthefly/impl/NestedNumericFieldComparatorSource.java:
```diff
@@ -0,0 +1,108 @@
+
+						BitSet parentDocs = docsProvider.parentDocs( context );
+						DocIdSetIterator childDocs = docsProvider.childDocs( context );
+						if ( parentDocs != null && childDocs != null ) {
```
Sure. I was thinking of doing something like that too.

fax4ever commented:
@yrodiere thanks.
I think I fulfilled the comments here.
If there were other changes to do, I would happy to handle them.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneNumericFieldSortBuilder.java:
```diff
@@ -39,17 +39,15 @@ public void buildAndContribute(LuceneSearchSortCollector collector) {
-			return new SortField( absoluteFieldPath, nestedFieldSort, order == SortOrder.DESC );
+			// propagate the comparator source, so that eventually it can receive the original query,
+			// that is required to nested document field on the fly sorting
+			this.nestedFieldSort = fieldComparatorSource;
```
I just remembered that `LuceneNumericFieldSortBuilder.buildAndContribute()` can be called concurrently from multiple threads. The mutating methods cannot, but `buildAndContribute` can. We could change that, but that would require adding yet another set of interfaces and classes, and, well... Not today.

Since `fieldComparatorSource` is mutable, this assignment could cause trouble.

Maybe you could just move this code to `buildAndContribute` and make `nestedFieldSort` a local variable?

fax4ever replied:
You're right. There is no reason to extract a field here.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneTextFieldSortBuilder.java:
```diff
@@ -41,28 +39,12 @@ public void buildAndContribute(LuceneSearchSortCollector collector) {
-			return nestedType();
+			// propagate the comparator source, so that eventually it can receive the original query,
+			// that is required to nested document field on the fly sorting
+			nestedFieldSort = fieldComparatorSource;
```
Same comment about `buildAndContribute` being called concurrently.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/impl/NestedNumericFieldComparatorSource.java:
```diff
@@ -0,0 +1,40 @@
+		fieldComparator = numericDomain.createNestedFieldComparator( fieldname, numHits, missingValue );
+
+		// workaround since we cannot extend a FieldComparator
+		return fieldComparator.getComparator();
```
1. `new FieldComparator.TermOrdValComparator( numHits, fieldname, sortMissingLast )` is invoked  during LuceneCollectors is being built.
2. `setOriginalParentQuery(Query luceneQuery)` that allows to have a `nestedDocsProvider` is invoked at the very beginning of `LuceneCollectors#collect`.
3. `getSortedDocValues(LeafReaderContext context, String field)` is invoked always on `LuceneCollectors#collect` a bit later than the previous one.

So the field `nestedDocProvider` is not set yet when `newComparator` method is called.

To handle that we should postpone the creation of the comparator (and the `TopFieldCollector`) from `LuceneCollectorsBuilder#build` to `LuceneCollectors#collect`, when we finally have the `Query luceneQuery`.

yrodiere replied:
I see.

Maybe you could move the call to `org.hibernate.search.backend.lucene.types.sort.comparatorsource.impl.LuceneFieldComparatorSource#setOriginalParentQuery` out of `org.hibernate.search.backend.lucene.search.extraction.impl.LuceneCollectors#collect` and to `org.hibernate.search.backend.lucene.search.query.impl.LuceneSearchQueryBuilder#build` instead? `definitiveLuceneQuery` is the exact same query that is currently passed to `LuceneCollectors#collect`.

I think this would allow you to get rid of the `nestedFieldSorts` contructor parameter in `LuceneSearchQueryRequestContext`, too, as the sorts will be completely initialized straight out of the query builder.

fax4ever replied:
Yes. You're right the query builder time is perfect. Done

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/impl/LuceneNumericFieldSortBuilder.java:
```diff
@@ -39,17 +39,15 @@ public void buildAndContribute(LuceneSearchSortCollector collector) {
-
-		SortField sortField = new SortField(
-				absoluteFieldPath,
-				numericDomain.getSortFieldType(),
```
You're right. We don't need it anymore.

fax4ever commented:
Addressed all comments, with the exception of this one: https://github.com/hibernate/hibernate-search/pull/2068#discussion_r322224342.

We could solve it, but we definitely need the final lucene query at `LuceneCollectors` build time (not just at invocation time). Do you think that is possible?
Or should we create TopFieldCollector at `LuceneCollectors` invocation time?
The latter sounds a bit strange...
Thanks!

fax4ever commented:
Running a full build: [job/PR-2068/19](https://ci.hibernate.org/job/hibernate-search/job/PR-2068/19)

fax4ever commented:
I need to add a commit to support the old nested sorting syntax of Elasticsearch 5 .

fax4ever commented:
Here we have a new build for all supported ES dialects: [job/PR-2068/22](https://ci.hibernate.org/job/hibernate-search/job/PR-2068/22)

fax4ever commented:
I added a test on a deep > 1 nested field. Full build: [job/PR-2068/24](https://ci.hibernate.org/job/hibernate-search/job/PR-2068/24)

fax4ever commented:
@yrodiere as usual you were right. Taking the flag from the `ElasticsearchJsonSyntaxHelper' is definitely much easier! You can find the jet another full build [here](https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2068/26).

yrodiere commented:
Merged! Thanks for all the work... and for putting up with me ;)
